Update Learn introduction page. (#1799)
graphql/graphql.github.io: src/pages/learn/index.mdx

import { Callout } from "nextra/components"

# Introduction to GraphQL

<p className="learn-subtitle">Learn about GraphQL, how it works, and how to use it</p>

> GraphQL is a query language for your API, and a server-side runtime for executing queries using a type system you define for your data. The [GraphQL specification](https://spec.graphql.org/) was open-sourced in 2015 and has since been implemented in a variety of programming languages. GraphQL also isn't tied to any specific database or storage engine and is instead backed by your existing code and data.

Reviewer: ```suggestion
GraphQL is a query language for your API, and a server-side runtime for executing queries using a type system you define for your data. The [GraphQL specification](https://spec.graphql.org/) was open-sourced in 2015 and has since been implemented in a variety of programming languages. GraphQL isn't tied to any specific database or storage engine—it is backed by your existing code and data.
```

> <Callout type="info">If you're already familiar with GraphQL and would like to read documentation on how to build a GraphQL service, then there are libraries to help you implement GraphQL in [many different languages](/community/tools-and-libraries/?tags=server). There are also many libraries available that allow [client applications to query existing GraphQL APIs](/community/tools-and-libraries/?tags=client).</Callout>

Reviewer 2: Should we mention about GraphQL.js being the official reference implementation maintained by the GraphQL foundation?

Author: @benjie Do you want to call out graphql.js specifically here?

Reviewer: I don't think we need to do that here, but our first mention of a "reference implementation" should hyperlink to GraphQL.js

> A GraphQL service is created by [defining types and their fields](/learn/schema/), and then writing functions that [provide data for each field](/learn/execution/). For example, a GraphQL service that tells you the name of a logged-in user might look like this:

Reviewer: This could be interpreted as you write a few functions that each apply to many fields; though there are functions like this (e.g. the default resolver), most fields get their own function. This slight rewording hopefully removes this ambiguity?

```suggestion
A GraphQL service is created by [defining types and their fields](/learn/schema/), and then writing a function for each field to [provide the required data](/learn/execution/). For example, a GraphQL service that tells you the name of a logged-in user might look like this:
```

```graphql
type Query {
  me: User
}

type User {
  name: String
}
```

Along with functions for each field on each type:

```js

> Query: {

Reviewer: This is no longer standalone JS code (`Query` here would be interpreted as a [label](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) which could be very confusing for readers new to JavaScript), and it ties into the `{[typeName]: {[fieldName]: resolver}}` pattern common in various tools across the JavaScript ecosystem but not actually supported by graphql.js natively and also not as portable to other programming languages.

`function Query_me(request)` has the advantage of being fairly generic pseudocode even if the reader does not know JavaScript.

Perhaps adding a comment before the function would be a suitable alternative solution:

```ts
// The resolver for the `me` field on the `Query` type, aka `Query.me`:
function Query_me(request) {
```

and later:

```ts
// The resolver for the `User.name` field:
function User_name(user) {
```

what do you think?

Author: I see what you mean with the label confusion. My intent here was to standardize these examples with the JS-ish examples that were already on the Execution page. But maybe a `function {Type}_{field} () {}` syntax would be best to use throughout the entire guide instead? Let me know what you think and I can adjust them.

```js
// Provide data for the `me` field on the `Query` type
function Query_me(obj, args, context) {
  return context.request.auth.user
}

// Provide data for the `name` field on the `User` type
function User_name(obj, args, context) {
  return context.db.getUserFullName(obj.id)
}
```

Reviewer: Urgh, yes, how inconsistent! I do like the structure of typeName/fieldName for resolver clarity but the invalid JS code is not helpful. Out of interest I fed the code from the execution page into ChatGPT to see what it made of it:

Converting the invalid JS code to Python:

https://chatgpt.com/share/671a2803-2338-8012-aaee-c7eb490b72ce (doesn't look useful)

Converting expanded valid JS code to Python:

https://chatgpt.com/share/671a2843-3f08-8012-83d7-8769c471ecd9 (looks much more reasonable to me).

Disclaimer: I've not written Python for ~8 years.

I think I'd stick with functions like you suggest above on this page, and then on the execution page if it makes sense to do the typeName/fieldName structure, explicitly introduce it (e.g. "for organizational reasons, you may wish to keep your resolvers in a structured object, for example ...")

Author: Sounds good. I updated the functions as discussed and moved the libraries callout in the latest commit so this should resolve all of the feedback for this page.

```js
// Provide data for the `me` field on the `Query` type

> function Query_me(obj, args, context, info) {
>   return context.request.auth.user

Reviewer: I wasn't sure whether to put `query` or `rootValue` or `source` or `parent` or `_` here; I went with `query` in the end for consistency with `user` below.

```suggestion
function Query_me(query, args, context, info) {
  return context.request.auth.user
```

}

// Provide data for the `name` field on the `User` type

> function User_name(obj, args, context, info) {
>   return context.db.getUserFullName(obj.id)

Reviewer: ```suggestion
function User_name(user, args, context, info) {
  return context.db.getUserFullName(user.id)
```

}
```

> In the example above, the function that provides data for the `me` field on the `Query` type uses information about the authenticated user who made the request, while the the `name` field on the `User` type is populated by using that user's ID to fetch their full name from an existing database.

Reviewer: I don't think "existing" is an important qualifier here, unless I'm missing something?

```suggestion
In the example above, the function that provides data for the `me` field on the `Query` type uses information about the authenticated user who made the request, while the the `name` field on the `User` type is populated by using that user's ID to fetch their full name from a database.
```

```json
{
  "data": {
    "me": {
      "name": "Luke Skywalker"
    }
  }
}
```

> With even a simple query, we can see some of the key features that make GraphQL so powerful. The client can make queries to the API that mirror the structure of the data that they need and then receive just that data in the expected shape with a single request—and without having to worry about what underlying data sources provided it.

Reviewer: ```suggestion
With even a simple query, we can see some of the key features that make GraphQL so powerful. The client can make queries to the API that mirror the structure of the data that they need and then receive just that data in the expected shape with a single request—and without having to worry about which underlying data sources provided it.
```

```json
{
  "data": {
    "me": {
      "name": "Luke Skywalker"
    }
  }
}
```

With even a simple query, we can see some of the key features that make GraphQL so powerful. The client can make queries to the API that mirror the structure of the data that they need and then receive just that data in the expected shape with a single request—and without having to worry about what underlying data sources provided it.

## Evolve your API without versioning

Client requirements change over time and GraphQL allows your API to evolve in response to those needs and without the overhead of managing different API versions. For example, if a new feature calls for more specific name values to be available, then the `User` type could be updated as follows:

```graphql
type User {
  fullName: String
  nickname: String
  name: String @deprecated(reason: "Use `fullName`.")
}
```

> GraphQL will continue to provide data as expected for the deprecated `name` field until it's determined that clients have updated their existing queries and the older field can be safely removed.

Reviewer: This edit clarifies that releasing an update to all of the clients is not sufficient - we also need to ensure that no existing (non-updated) clients remain - e.g. mobile apps from years ago. It also makes it clear that GraphQL itself does not make any determination that the field is no longer used.

```suggestion
Client tooling will encourage developers to use the new fields and remove usage of the deprecated `name` field. The field can be removed once it is determined it is no longer used; in the meantime GraphQL will continue to provide its data as expected.
```

## Try it out!

The best way to learn GraphQL is to start writing queries. The query editors used throughout this guide are **interactive**, so try adding an `id` and `appearsIn` fields to the `hero` object to see an updated result:

```graphql
# { "graphiql": true }
{
  hero {
    name
    # add additional fields here!
  }
}
```

<Callout type="info">
The examples in this guide are based on [a modified version of the SWAPI GraphQL schema](https://github.com/graphql/graphql.github.io/blob/source/src/components/marked/swapi-schema.tsx). Because these queries are designed for illustrative purposes, they will not run on the full version of the SWAPI GraphQL API due to differences between the two schemas. [You can try the full version of the API here.](https://graphql.org/swapi-graphql/)
</Callout>

## Next steps

Now that you know some key GraphQL concepts, you're ready to learn about all of the different aspects of its [type system](/schema/).

> If you're already familiar with GraphQL and would like to read documentation on how to build a GraphQL service, then there are libraries to help you implement GraphQL in [many different languages](/community/tools-and-libraries/?tags=server). There are also many libraries available that allow [client applications to query existing GraphQL APIs](/community/tools-and-libraries/?tags=client).

Reviewer: If someone is already familiar with GraphQL it seems unlikely they would read this far - perhaps this should be in a callout at the top of the page? I'm not super familiar with Nextra, but in Docusaurus I'd use an "info" [admonition](https://docusaurus.io/docs/markdown-features/admonitions).

Author: Ok, I'll move this back to the top.
